LFS-795, LFS-366: Refactor Date questions and fix DateTime

acrowthe opened the pull request:
Testing branch available [here](https://github.com/ccmbioinfo/lfs/tree/LFS-795-test). Testing branch adds a new form in the cardiac_rehab runmode (1 Date Test) so many date options and formats can be tested.

Change DateQuestion from storing a date to a string representing the date
- Remove separate "displayFormat" and "dateFormat" options
- Fixes date+time and date durations
- Improves output format options for patient charts and (future) timeline

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions
Change DateQuestion from storing a date to a string representing the date
- Remove separate "displayFormat" and "dateFormat" options
- Fixes date+time and date durations
- Improves output format options for patient charts and (future) timeline

acrowthe committed:
LFS-795: : Refactor Date question
- Add a fix so days are formatted to numbers, not text representing day of week

sashaandjic commented:
no validation for a year field
![image](https://user-images.githubusercontent.com/23243244/103741800-bba26380-4fc7-11eb-8a98-a5e6a77a5b67.png)

I am unable to edit seconds - they turn into AM or PM

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
Split year, month and date questions into separate components

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionFull.jsx:
```diff
@@ -0,0 +1,194 @@
+
+  let stripTimeZone = (dateString) => {
+    // Remove the time zone (eg. "-5:00") from the end of a sling provided date string
+    return dateString.replace(/-[0-9]{2}:[0-9]{2}$/gm, '');
```
The timezone could also be +hh:mm

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionFull.jsx:
```diff
@@ -0,0 +1,194 @@
+    ? DateQuestionUtilities.amendMoment(stripTimeZone(startValues[1]), DateQuestionUtilities.slingDateFormat)
+    : null
+  );
+  const upperLimitMoment = upperLimit ? DateQuestionUtilities.amendMoment(upperLimit, DateQuestionUtilities.slingDateFormat) : null;
```
This is repeated several times. Let's make it shorter:
* `DateQuestionUtilities.amendMoment` should return `null` if the first parameter is `null`
* If the second (format) parameter is missing, ` DateQuestionUtilities.slingDateFormat` should be used by default
```suggestion
  const upperLimitMoment = DateQuestionUtilities.amendMoment(upperLimit);
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionFull.jsx:
```diff
@@ -0,0 +1,194 @@
+
+  let startValues = existingAnswer && existingAnswer[1].value || "";
+
+  let stripTimeZone = (dateString) => {
```
Can this be moved to the utilities?

sashaandjic commented:
I can still enter 1, 2 or 3 digit year in the Year field of the testing form

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
- Change year question lower limit to 1000 instead of 0

sashaandjic reviewed (approved):
tested

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
Address code review comments
- Clean up amendMoment calls
- Fix strip time zone logic and move to utilities

sashaandjic commented:
When using date/time picker I can set an invalid date range
![image](https://user-images.githubusercontent.com/23243244/105092914-e54f9600-5a6f-11eb-9b3d-164a63bbaea9.png)

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
Fix issues with interval dates behaving incorrectly
- Full dates not requiring valid end date
- Month dates reseting start to match end if end date is present

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
Move full date question on change logic to on blur
- Fixes issue where user could not manually type a change to an interval end
  if a partially entered value would get overwritten by the start

sashaandjic commented:
Problem found:
Trying to set **ITSEA** flag in carefkids' **0-5NDVariables Data** form  causes form to go blank.
![image](https://user-images.githubusercontent.com/23243244/105531091-ca229780-5cb6-11eb-813a-b347fda66d83.png)
![setting ITSEA](https://user-images.githubusercontent.com/23243244/105531293-1968c800-5cb7-11eb-8c26-45b31ac819b3.png)

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
- Fix errors when getting the date type for an undefined date format string

sashaandjic reviewed (approved):
tested

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
Fix merge issue

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionMonth.jsx:
```diff
@@ -0,0 +1,233 @@
+DateQuestionMonth.propTypes = DateQuestionUtilities.PROP_TYPES;
+
+const StyledDateQuestion = withStyles(QuestionnaireStyle)(DateQuestionMonth);
+export default StyledDateQuestion;
```
All these new files have the same default export component name, `StyledDateQuestion`. Will this cause any issues? I would prefer that they have different names, `StyledDateQuestionMonth`, `StyledDateQuestionFull`, `StyledDatedQuestionYear`.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionMonth.jsx:
```diff
@@ -0,0 +1,233 @@
+  {
+    return [StyledDateQuestion, 60];
+  }
+});
```
Please add newlines at the end of files.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -22,7 +22,7 @@ import { Link, useLocation, withRouter } from 'react-router-dom';
 import moment from "moment";
 
-import { formatDateAnswer } from "./DateQuestion.jsx";
+import DateQuestionUtilities from "./DateQuestionUtilities.jsx";
```
This isn't used at all, just remove it.

```suggestion
```

sdumitriu commented on modules/data-entry/src/main/frontend/webpack.config.js:
```diff
@@ -23,7 +23,7 @@ module.exports = {
     [module_name + 'questionnairesIcon']: '@material-ui/icons/Assignment.js',
     [module_name + 'formsIcon']: '@material-ui/icons/Description.js',
-    [module_name + 'dateQuestion']: './src/questionnaire/DateQuestion.jsx',
+    [module_name + 'dateQuestionUtilities']: './src/questionnaire/DateQuestionUtilities.jsx',
```
Does this need to be an entry at all? I have a feeling it doesn't.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/FilterComponents/DateFilter.jsx:
```diff
@@ -43,8 +43,9 @@ const DateFilter = forwardRef((props, ref) => {
-  let isMonth = MONTH_FORMATS.includes(dateFormat);
-  let isDate = DATE_FORMATS.includes(dateFormat);
+  let dateType = DateQuestionUtilities.getDateType(dateFormat);
+  let isMonth = (dateType === DateQuestionUtilities.MONTH_DATE_TYPE);
```
I would like to have methods for checking the format instead of direct comparison to constants, i.e. `let isMonth = DateQuestionUtilities.isMonthFormat(dateFormat)`, but that can be done later.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionMonth.jsx:
```diff
@@ -0,0 +1,233 @@
+
+import { TextField, Typography, withStyles } from "@material-ui/core";
+
+import PropTypes from "prop-types";
```
```suggestion
```

acrowthe committed:
LFS-795, LFS-366: Refactor Date questions and fix DateTime
Address review comments
- Clean up Javascript imports
- Adjust exported question names to avoid duplication
- Add trailing newlines to files
- Clean up webpack exports
